### PR TITLE
Add Kubernetes Dashboard, fix README, various small fixes

### PR DESCRIPTION
**Pull request by diemonster**
This PR adds the Kubernetes Dashboard using [the tutorial instructions ](https://docs.aws.amazon.com/eks/latest/userguide/dashboard-tutorial.html) provided by Amazon.

It also:

- updates the README to remove some ambiguity

- adds `etcd-operator` to the root-level Makefile

- removes the root Makefile's `uninstall` target. It didn't work, and it's not necessary to run before executing `terraform destroy clustername`

...and some more small tweaks and fixes.
 
Closes https://github.com/quintilesims/production_eks_cluster/issues/19 https://github.com/quintilesims/production_eks_cluster/issues/21

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,18 +1,17 @@
+# Set PATH to include local dir for locally downloaded binaries.
+export PATH := .:$(PATH)
+
 KUBECONFIG?="~/.kube/config.eks"
 
-install:
+install: check-env
```
Comment (nehayward) [inline Makefile]: No `check-env` target

File: Makefile [hunk 2]
```diff
@@ -1,18 +1,17 @@
...
 	cp -f kubeconfig_* $(KUBECONFIG)
 	kubectl apply -f config-map-aws-auth_*.yaml
 	$(MAKE) -C addons/helm install
+	$(MAKE) -C addons/dashboard install
 	$(MAKE) -C addons/logging install
 	$(MAKE) -C addons/prometheus install
+	$(MAKE) -C addons/etcd install
 	$(MAKE) -C addons/sso install
 
-uninstall:
```
Comment (nehayward) [inline Makefile]: This shouldn't be removed either.

File: addons/dashboard/get-token.sh [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#!/usr/bin/env bash
```
Comment (nehayward) [inline addons/dashboard/get-token.sh]: Should we include this in the makefile instead? Seems more consistent with the rest of the add ons.

Comment (diemonster) [reply addons/dashboard/get-token.sh]: I'm not a fan of placing BASH commands in Makefiles; in this case, the substitution won't work in a Makefile without suppling a `SHELL` param, which I've had mixed results with.

File: Makefile (hunk 2 above)
Comment (diemonster) [reply Makefile]: I would argue that the `uninstall` target in the root-level Makefile is not necessary. For one, several of the addons don't actually uninstall. But beyond that, `terraform destroy` works fine without removing all (or any) of the addons.

File: addons/dashboard/get-token.sh (hunk 3 above)
Comment (nehayward) [reply addons/dashboard/get-token.sh]: Okay that's a solid argument for me and I've run into that before it muddies the Makefile.

File: Makefile (hunk 2 above)
Comment (nehayward) [reply Makefile]: Yup as long as terraform destroy doesn't leave hangin resources, which was a problem with some of the add ons.

File: Makefile [hunk 4]
```diff
@@ -1,18 +1,17 @@
+# Set PATH to include local dir for locally downloaded binaries.
+export PATH := .:$(PATH)
```
Comment (zpatrick) [inline Makefile]: Not sure why this is here? The top-level README is pretty clear about any required binaries being in the PATH, and if there are any binaries in this directory, they should just be ran with `./`

File: Makefile (hunk 2 above)
Comment (zpatrick) [reply Makefile]: > For one, several of the addons don't actually uninstall

That shouldn't be the case; and it is a bug if what you are saying is true. Which addons don't uninstall?

> terraform destroy works fine without removing all (or any) of the addons.

That's not true - at least one of the addons (can't remember which one off the top of my head) creates an ELB that will prevent `terraform destroy` from succeeding. Also, the prometheus addon creates a new storage class in K8, which creates EBS volumes in AWS, which will be left around if they aren't deleted.

File: addons/dashboard/Makefile [hunk 5]
```diff
@@ -0,0 +1,12 @@
+.PHONY: install update
```
Comment (zpatrick) [inline addons/dashboard/Makefile]: This makefile needs an `uninstall` target.

File: addons/dashboard/Makefile [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+install:
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/dashboard/master/src/deploy/recommended/kubernetes-dashboard.yaml
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/heapster/master/deploy/kube-config/influxdb/heapster.yaml
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/heapster/master/deploy/kube-config/influxdb/influxdb.yaml
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/heapster/master/deploy/kube-config/rbac/heapster-rbac.yaml
+	kubectl apply -f eks-admin-service-account.yaml
+	kubectl apply -f eks-admin-cluster-role-binding.yaml
+	@./get-token.sh
+
+update:
```
Comment (zpatrick) [inline addons/dashboard/Makefile]: this looks more like a delete than an update.

File: addons/dashboard/README.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
+## Enabling the Kubernetes Dashboard
+
+Run the Makefile we've included to configure the [EKS Kubernetes Dashboard](https://docs.aws.amazon.com/eks/latest/userguide/dashboard-tutorial.html). 
+
+```
+make dashboard
```
Comment (zpatrick) [inline addons/dashboard/README.md]: should be `make install`

File: addons/etcd-operator/.gitignore [hunk 8]
```diff
@@ -0,0 +1,3 @@
+# Role bindings used by Makefile
```
Comment (zpatrick) [inline addons/etcd-operator/.gitignore]: why are these being ignored? they are in the repo

File: Makefile (hunk 2 above)
Comment (diemonster) [reply Makefile]: SSO creates an ELB that prevents `terraform destroy` from succeeding, but I think we should address that in another place than this root-level Makefile. I actually thought `sso` was going to be moved to it's own repo and not be made part of `addons` by default.

Could we add resources to terraform state if/when they come from external places? Is there a clean way to do that? 

Anyway, output from my last uninstall attempt:
```
make uninstall
/Library/Developer/CommandLineTools/usr/bin/make -C addons/helm uninstall
kubectl delete -f role_binding.yaml
clusterrolebinding.rbac.authorization.k8s.io "tiller" deleted
kubectl delete -f service_account.yaml
serviceaccount "tiller" deleted
/Library/Developer/CommandLineTools/usr/bin/make -C addons/logging uninstall
helm del --purge fluentd
Error: Unauthorized
```

File: Makefile (hunk 4 above)
Comment (diemonster) [reply Makefile]: I don't feel super strongly about this one, but adding the current execution path to `$PATH` during `make install` is a another measure that helps when you have version mis-matches locally of either `kubectl` or `aws-iam-authenticator`. I've found a lot of breaking changes between versions w both of these utilities, and it's been something that several people (including myself) have gotten stuck on.

File: addons/etcd-operator/.gitignore (hunk 8 above)
Comment (diemonster) [reply addons/etcd-operator/.gitignore]: No, neither of these files are in the repo. They great created as part of `make install`.

File: addons/dashboard/Makefile (hunk 6 above)
Comment (diemonster) [reply addons/dashboard/Makefile]: https://github.com/kubernetes/dashboard/wiki/Installation#update

File: Makefile (hunk 2 above)
Comment (zpatrick) [reply Makefile]: > but I think we should address that in another place than this root-level Makefile

I'm not following your logic here - if `install` is fine in this scope, then why wouldn't `uninstall` be? 

>  I actually thought sso was going to be moved to it's own repo and not be made part of addons by default.

The source will live in its own repo, but like prometheus or dashboard addons, we will have a directory for it in this repo that contains a Makefile and the required yaml files. 

> Could we add resources to terraform state if/when they come from external places? Is there a clean way to do that?

You can use `terraform import`, but that's not a good idea in this context for multiple reasons. 
Long story short, we decided using Makefiles with `install` and `uninstall` targets allows us to programmatically create/destroy resources in a sane way without jumping through hoops to get everything managed by terraform. 

> helm del --purge fluentd
Error: Unauthorized

That looks like helm isn't configured correctly on your machine, or there's a bug with that addon. I think you should create an issue for that and cc @nehayward since he created that addon.

File: addons/dashboard/Makefile (hunk 6 above)
Comment (zpatrick) [reply addons/dashboard/Makefile]: that example is titled "Delete all Dashboard pods (assuming that Dashboard is deployed in kube-system namespace)"

Comment (diemonster) [reply addons/dashboard/Makefile]: > Once installed, the deployment is not automatically updated. In order to update it you need to delete the deployment's pods and wait for it to be recreated. After recreation, it should use the latest image.

> Delete all Dashboard pods (assuming that Dashboard is deployed in kube-system namespace):

File: addons/dashboard/Makefile (hunk 5 above)
Comment (diemonster) [reply addons/dashboard/Makefile]: I agree, I'll add this. It's kind of a pain in the butt and `terraform destroy` works regardless of whether it's installed... but it's still the Right Thing To Do

File: addons/dashboard/Makefile (hunk 6 above)
Comment (zpatrick) [reply addons/dashboard/Makefile]: We don't provide `update` for any other addons, I don't know why this one should be an exception?

Jumping more into personal preference/philosophy here, but I think it's best to give targets names that describe what the commands are literally doing, and then, if desired, one can create higher level targets by mixing-and-matching the literal targets, e.g.:
```make
create: 
    kubectl create ...

delete:
    kubectl delete ...

# there is no literal update - just delete and re-create
update: delete create
```

File: Makefile (hunk 4 above)
Comment (zpatrick) [reply Makefile]: I think if we lock down the eks cluster module to a specific version, we could potentially fix a lot of those issues. We could revert back to a version that didn't require the `aws-iam-authenticator` binary name change.

File: addons/dashboard/Makefile (hunk 6 above)
Comment (diemonster) [reply addons/dashboard/Makefile]: The other addons, with the exception of `helm` itself, are managed by `helm`. In other words, `helm` provides an easy path for upgrades and updates. Such is not the case for the actual dashboard components.

Your other question might be: "why not use helm to install the Dashboard?". Answer to that is that there are some EKS caveats, and being unable to use helm to install the dashboard is one of them. For now.

I like your idea on the [nit] though, that's a clearer option for sure.

File: Makefile (hunk 4 above)
Comment (diemonster) [reply Makefile]: it's mostly an issue for anyone that has been using `kubectl` in the last year without diligently updating their binary. Only the latest release of `kubectl` supports EKS (at any version), but unfortunately the error message when using an older version is that you get prompted for authentication. 

The same is generally true for `helm` and `aws-iam-authenticator`: unhelpful errors that seem unrelated to the version of the client. I can't really see how fixing to a version of EKS would solve any of those issues.

File: addons/etcd-operator/Makefile [hunk 9]
```diff
@@ -1,6 +1,7 @@
 install:
-	helm install stable/etcd-operator --name etcd-operator
-	kubectl create -f etcd-cluster.yaml
+	helm upgrade --install etcd-operator stable/etcd-operator
+	@sleep 10
```
Comment (nehayward) [inline addons/etcd-operator/Makefile]: Is this pretty consistent in timing? Or should we add a loop?

File: Makefile [hunk 10]
```diff
@@ -1,18 +1,35 @@
...
-	cp -f kubeconfig_* $(KUBECONFIG)
+
+tf_destroy:
+	terraform destroy
+
+set_kubeconfig:
+	$(shell cp -rf kubeconfig_* $(KUBECONFIG))
 	kubectl apply -f config-map-aws-auth_*.yaml
+
+install: | tf_init tf_apply set_kubeconfig
```
Comment (nehayward) [inline Makefile]: What's does`|` do?

File: Makefile [hunk 11]
```diff
@@ -1,18 +1,35 @@
...
+
+tf_apply:
 	terraform apply
-	cp -f kubeconfig_* $(KUBECONFIG)
+
+tf_destroy:
+	terraform destroy
+
+set_kubeconfig:
+	$(shell cp -rf kubeconfig_* $(KUBECONFIG))
```
Comment (diemonster) [inline Makefile]: I should call out that this still appears to be broken, or at the very least, not portable. Commit incoming to fix this.

File: README.md [hunk 12]
```diff
@@ -5,8 +5,13 @@ This example shows how one can integrate the [AWS VPC](https://registry.terrafor
 
 * [AWS Authentication](https://www.terraform.io/docs/providers/aws/index.html#authentication) setup on your local machine
 * [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/) (>= v1.10.4) installed and in your local machine's PATH
-* [Heptio Authenticator](https://github.com/heptio/authenticator) (>= v0.3.0) installed and in your local machine's PATH
-* [Helm](https://github.com/kubernetes/helm) (>=v2.9.0) installed and in your local machine's PATH
+* [Heptio Authenticator](https://github.com/heptio/authenticator) (>= v0.3.0) installed and in your local machine's PATH. Currently, this binary needs to be renamed to `aws-iam-authenticator` after being placed in your PATH.
+* [Helm](https://github.com/kubernetes/helm) (>=v2.9.0) installed and in your local machine's PATH.
+* The following environment variables must be set:
+  * `KUBECONFIG` - this tells `kubectl` where your configuration should reside. We recommend setting it to a value like `~/.kube/config.eks`.
+  * `AWS_ACCESS_KEY_ID` - your AWS access key.
```
Comment (tlake) [inline README.md]: Is this (and similar) env var needed for anything beyond terraform? Currently, following the readme as is results in terraform prompting me for `var.aws_access_key`, `var.aws_secret_key`, and `var.cluster_name`. Is this intentional? Should we be telling users to set `TF_VAR_aws_access_key`, etc. instead of `AWS_ACCESS_KEY_ID`? Possibly in addition to it?

Comment (diemonster) [reply README.md]: All three of the mentioned env vars MUST be set in order for the logging addon to be installed. 

The Terraform prompts, on the other hand, are annoying but as long as you supply them, you'll still be able to proceed.

Comment (diemonster) [reply README.md]: p.s. i also find it really aggravating to have to do something like this:
```
export TF_VAR_aws_access_key=
export TF_VAR_aws_secret_key=
```

File: addons/dashboard/Makefile [hunk 13]
```diff
@@ -0,0 +1,31 @@
+install:
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/dashboard/master/src/deploy/recommended/kubernetes-dashboard.yaml
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/heapster/master/deploy/kube-config/influxdb/heapster.yaml
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/heapster/master/deploy/kube-config/influxdb/influxdb.yaml
+	kubectl apply -f https://raw.githubusercontent.com/kubernetes/heapster/master/deploy/kube-config/rbac/heapster-rbac.yaml
```
[the remark this answers is not in the document]
Comment (diemonster) [reply addons/dashboard/Makefile]: Not sure what you're referring to --- the `uninstall` portion refers to the entities created when it deletes them, the URLs are not needed.

[the remark this answers is not in the document]
Comment (nehayward) [reply addons/dashboard/Makefile]: Yes they aren't needed, but it may be easier to follow with this instead. It's a nit
```kubectl delete -f https://raw.githubusercontent.com/kubernetes/heapster/master/deploy/kube-config/rbac/heapster-rbac.yaml```

File: README.md (hunk 12 above)
Comment (tlake) [reply README.md]: Heh, I find it much more preferable to do that once than to copy&paste each env var into terraform for every plan/apply/destroy. :p

File: README.md [hunk 14]
```diff
@@ -5,8 +5,13 @@ This example shows how one can integrate the [AWS VPC](https://registry.terrafor
...
+  * `AWS_SECRET_ACCESS_KEY` - your AWS secret access key.
+  * `LOG_GROUP_NAME` - the name for your cluster's log group in AWS Cloudwatch.
 
 ## Walkthrough
```
Comment (tlake) [inline README.md]: Going through it, it seems like maybe this walkthrough section is outdated. Like, it seems like `make install` is the preferred entrypoint, not `terraform init|plan|apply`.

File: addons/etcd-operator/Makefile (hunk 9 above)
Comment (diemonster) [reply addons/etcd-operator/Makefile]: If you want to write a loop, go for it

File: README.md (hunk 12 above)
Comment (diemonster) [reply README.md]: I think you misunderstood me --- I think it's a bit silly that the same variable has to be set twice. But the alternative of having to type the access key and secret key every time is far worse.

Comment (tlake) [reply README.md]: Ah, I did. I getcha now.

File: Makefile [hunk 15]
```diff
@@ -1,18 +1,33 @@
...
+# WARNING: uninstall is not idempotent 
+# https://github.com/quintilesims/production_eks_cluster/issues/20
 uninstall:
 	$(MAKE) -C addons/sso uninstall
 	$(MAKE) -C addons/prometheus uninstall
 	$(MAKE) -C addons/logging uninstall
+	$(MAKE) -C addons/etcd-operator uninstall
 	$(MAKE) -C addons/helm uninstall
+	$(MAKE) -C addons/dashboard uninstall
+	tf_destroy
```
Comment (tlake) [inline Makefile]: This throws an error at the last step because `tf_destroy` isn't a file or command. I don't think you can specify a make target here like this. We might want to do something like this:

```
uninstall_addons:
	$(MAKE) -C addons/sso uninstall
	. . . . .

uninstall: uninstall_addons tf_destroy
```

File: README.md (hunk 12 above)
Comment (tlake) [reply README.md]: Do you think there's value in passing the environment variables as `TF_VAR_` env vars within the makefile to the terraform targets?

Comment (diemonster) [reply README.md]: Sure, if you want to do that, file an issue and open a separate PR.

File: README.md [hunk 16]
```diff
@@ -3,53 +3,33 @@ This example shows how one can integrate the [AWS VPC](https://registry.terrafor
...
-Run the following command(s) from this directory:
-```console
-$ terraform destroy
-...
-
-Destroy complete! Resources: 47 destroyed.
+make uninstall
 ```
+
+*NOTE* `make uninstall` is not idempotent. If you encounter any errors during the uninstall process, you must manually uninstall any remaining addons, and then run `terraform destroy` to remove the remaining AWS resources.
```
Comment (zpatrick) [inline README.md]: with the `--ignore-not-found` flags, is this still true?

Comment (diemonster) [reply README.md]: Yes. For example: https://github.com/quintilesims/production_eks_cluster/issues/20
